app/gui2/src/components/GraphEditor/clipboard.ts

```diff
@@ -14,107 +21,148 @@ interface ClipboardData {
-          }
-        }
+export function useGraphEditorClipboard(
+  nodeSelection: GraphSelection,
+  graphNavigator: NavigatorComposable,
+) {
+  const graphStore = useGraphStore()
 
-        if (type === 'text/plain') {
-          const blob = await clipboardItem.getType(type)
-          const fallbackExpression = await blob.text()
-          const fallbackNode = { expression: fallbackExpression, metadata: undefined } as CopiedNode
-          fallback = { nodes: [fallbackNode] } as ClipboardData
-        }
-      }
+  /** Copy the content of the selected node to the clipboard. */
+  function copySelectionToClipboard() {
+    const nodes = new Array<Node>()
+    for (const id of nodeSelection.selected) {
+      const node = graphStore.db.nodeIdToNode.get(id)
+      if (!node) continue
+      nodes.push(node)
     }
-    return fallback
+    if (!nodes.length) return
+    getClipboard()
+      .write(nodesToClipboardData(nodes))
+      .catch((error: any) => console.error(`Failed to write to clipboard: ${error}`))
   }
 
-  /// Read the clipboard and if it contains valid data, create a node from the content.
-  async function readNodeFromClipboard() {
-    const clipboardData = await retrieveDataFromClipboard()
-    const copiedNode = clipboardData?.nodes[0]
-    if (!copiedNode) {
+  /** Read the clipboard and if it contains valid data, create nodes from the content. */
+  async function createNodesFromClipboard() {
+    const clipboardItems = await getClipboard().read()
+    const clipboardData = await nodesFromClipboardContent(clipboardItems)
+    if (!clipboardData.length) {
       console.warn('No valid node in clipboard.')
       return
     }
-    if (copiedNode.expression == null) {
-      console.warn('No valid expression in clipboard.')
-    }
-    graphStore.createNode(
-      graphNavigator.sceneMousePos ?? Vec2.Zero,
-      copiedNode.expression,
-      copiedNode.metadata,
-    )
-  }
-
-  async function readNodeFromExcelClipboard(
-    htmlContent: string,
-    clipboardItem: ClipboardItem,
-  ): Promise<ClipboardData | undefined> {
-    // Check we have a valid HTML table
-    // If it is Excel, we should have a plain-text version of the table with tab separators.
-    if (
-      clipboardItem.types.includes('text/plain') &&
-      htmlContent.startsWith('<table ') &&
-      htmlContent.endsWith('</table>')
-    ) {
-      const textData = await clipboardItem.getType('text/plain')
-      const text = await textData.text()
-      const payload = JSON.stringify(text).replaceAll(/^"|"$/g, '').replaceAll("'", "\\'")
-      const expression = `'${payload}'.to Table`
-      return { nodes: [{ expression: expression, metadata: undefined }] } as ClipboardData
+    for (const copiedNode of clipboardData) {
+      const { expression, documentation, visualization, colorOverride } = copiedNode
+      graphStore.createNode(
+        (clipboardData.length === 1 ? graphNavigator.sceneMousePos : null) ?? Vec2.Zero,
```

Comment:
Positioning of multiple nodes is currently rudimentary; this will be fixed with the new API that will be introduced for #9703.